### PR TITLE
[Core][Geometry] Add Assign and Calculate interfaces

### PR DESCRIPTION
**Pull request by tteschemacher**
yo,

welcome back with a beautiful PR after the summer break.

**Description**
This PR adds the interfaces for Set and Calculate to the geometry. They are similar as the ones from Elements and Conditions. This allows dynamic interfaces, especially with more sophisticated geometries. Like this we do not have to add new interfaces, but can access to all internal variables and compute upon them.

Normally it would be Get and Set. However, to be consistent it is Calculate and Set.

### CODE REVIEW

**Comment by sunethwarna** [comment]
Hi, I am just curious, why don't we support array_1d<double, 4> and array_1d<double, 9> in these interfaces? Ofcourse, its not for this PR, even `Element` interface does not have them.

**Comment by philbucher** [comment]
> Hi, I am just curious, why don't we support array_1d<double, 4> and array_1d<double, 9> in these interfaces? Ofcourse, its not for this PR, even `Element` interface does not have them.

Don't wake the sleeping dogs :see_no_evil: 
Basically bcs none needed it so far

File: kratos/geometries/geometry.h [hunk 1]
```diff
@@ -634,6 +634,99 @@ class Geometry
...
+
+    /// Set with array_1d<double, 3>
+    virtual void Set(
+        const Variable<array_1d<double, 3>>& rVariable,
+        array_1d<double, 3>& Output) {}
+
+    /// Set with array_1d<double, 6>
+    virtual void Set(
+        const Variable<array_1d<double, 6>>& rVariable,
+        array_1d<double, 6>& Output) {}
```
Comment (philbucher) [inline kratos/geometries/geometry.h]:
```suggestion
    /// Set with array_1d<double, 3>
    virtual void Set(
        const Variable<array_1d<double, 3>>& rVariable,
        array_1d<double, 3>& Output) {}
```
I would keep the extension separate

File: kratos/geometries/geometry.h [hunk 2]
```diff
@@ -634,6 +634,99 @@ class Geometry
...
+
+    /// Set with int
+    virtual void Set(
+        const Variable<int>& rVariable,
+        int& Output) {}
+
+    /// Set with double
+    virtual void Set(
+        const Variable<double>& rVariable,
+        double& Output) {}
```
Comment (philbucher) [inline kratos/geometries/geometry.h]:
```suggestion
        const double Input {}
```
this waythe usage is clearer

File: kratos/geometries/geometry.h [hunk 3]
```diff
@@ -634,6 +634,99 @@ class Geometry
...
+
+    /// Calculate with Vector
+    virtual void Calculate(
+        const Variable<Vector>& rVariable,
+        Vector& Output) {}
+
+    /// Calculate with Matrix
+    virtual void Calculate(
+        const Variable<Matrix>& rVariable,
+        Matrix& Output) {}
```
Comment (philbucher) [inline kratos/geometries/geometry.h]:
```suggestion
        Matrix& rOutput) {}
```

Comment (tteschemacher) [reply kratos/geometries/geometry.h]: yo, changed it.

File: kratos/geometries/geometry.h (hunk 2 above)
Comment (tteschemacher) [reply kratos/geometries/geometry.h]: definitely. Thanks. I changed it

File: kratos/geometries/geometry.h (hunk 1 above)
Comment (tteschemacher) [reply kratos/geometries/geometry.h]: To be honest, right now, I am explicitly after the array_1d<double, 2>. I'd actually like to keep it.

**Review by tteschemacher — COMMENTED** [review]
Thanks for the comments. I changed the namings, but I would prefer to keep the extended functions.

**Comment by tteschemacher** [comment]
@KratosMultiphysics/technical-committee 

Just asking, was there any decision taken, yet? I know that this is a bigger change however, I think the code is pretty obvious as it is. Thanks!

**Comment by philbucher** [comment]
you didn't add it to the KTC project :)

**Comment by tteschemacher** [comment]
> you didn't add it to the KTC project :)

I didn't want to go that big. I was hoping for a quiet approval :P

File: kratos/python/add_geometries_to_python.cpp [hunk 4]
```diff
@@ -163,6 +185,24 @@ void  AddGeometriesToPython(pybind11::module& m)
     .def("Length",&GeometryType::Length)
     .def("Area",&GeometryType::Area)
     .def("Volume",&GeometryType::Volume)
+    // Calculate
```
Comment (rubenzorrilla) [inline kratos/python/add_geometries_to_python.cpp]:
```suggestion
    // Set
```

**Review by rubenzorrilla — COMMENTED** [review]
I've two comments:
- Why in the `Set` methods we are passing `arrays_1d<double,X>`, `Vector` and `Matrix` by copy?
- `array_1d<double,2>`? 🙈

**Comment by tteschemacher** [comment]
>     * Why in the `Set` methods we are passing `arrays_1d<double,X>`, `Vector` and `Matrix` by copy?
I changed it.
>     * `array_1d<double,2>`? 🙈
That's the vibe... Soon you'll love it, too.

**Review by pooyan-dadvand — COMMENTED** [review]
@KratosMultiphysics/technical-committee we see this addition to the interface as a way to reduce newer methods in the Geometry and we agree to their addition following these guidelines:
* Set should be renamed to SetValue to be consistent with elements and conditions
* Only the types that will be used now should be added to the interface (`double`, `array_1d<double,3>` and `Vector`) and the rest of the types should be included when we have a use case.

**Comment by tteschemacher** [comment]
Hey @pooyan-dadvand , thanks for your comments.

> @KratosMultiphysics/technical-committee we see this addition to the interface as a way to reduce newer methods in the Geometry and we agree to their addition following these guidelines:
>     * Set should be renamed to SetValue to be consistent with elements and conditions
As the geometry is holding now the data container it has already the function SetValue, which is non-virtual as it deals with the data container. Thus, I chose to name it Set. It could be also SetOnGeometry, however, then it would be GetGeometry().SetOnGeometry(). There's a lot of redundancy Geometry to my opinion. However, I am open for discussion.

> Only the types that will be used now should be added to the interface (double, array_1d<double,3> and Vector) and the rest of the types should be included when we have a use case.
I am after the array1_d<double, 2>, it doesn't really harm having the full spectrum does it? In the other way round it would either produce another overhead and additionally developers would come up with dodgy ways to avoid adding the complete interfaces.

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee sees two different options with their own pros and cons:

1. Having two sets of methods, one non virtual and templated (currently SetValue) and other one virtual (Set) as is proposed in this PR
2. Only having the virtual methods and eliminating the templated ones. 

Option 1 has the advantage of performance in using non virtual methods and you may store any kind of variables in it. But the outcome is undefined if someone changes some intrinsic variable of the geometry (like radius or curvature) by the base class method without passing through the correct derived one.

While option 2 avoids the problem of option 1 but misses the advantages mentioned before. For example, it may cause some performance overhead due to the virtual table call.

So any comments and suggestions are welcomed. @KratosMultiphysics/implementation-committee

**Comment by tteschemacher** [comment]
I'm for option 1. It's more in the way as it is already. And then we could add it as well to the Element/Condition. Like this we have a continuous solution. SetValue is container and non virtual, Set is virtual.

To my opinion option 2 had the problem that you would always need to refer to the base class which can be forgotten easily.

**Comment by pooyan-dadvand** [comment]
@tteschemacher We don't understood your comment about the option 2. Would you please elaborate e little bit more?

@KratosMultiphysics/implementation-committee any comment?

**Comment by pooyan-dadvand** [comment]
@tteschemacher would you please elaborate more the second point?

**Comment by philbucher** [comment]
> @tteschemacher would you please elaborate more the second point?

ping @tteschemacher

**Comment by tteschemacher** [comment]
I think I also don't really understand the question.

So the case is the following: we have the function interfaces GetValue and SetValue which are comunicating with the data value container. This is a dynamic memory. It is very widely used throughout KRATOS, as it contains now directly the information from the Elements or Conditions, accordingly. Touching them would not be an option.

Now with this PR I want to have an interface which can communicate with the static memory of the geometries. Examples here are: #7552, #7811, #8040 and #8085. We cannot use the dynamic access, here.

Now I am not 100% sure what you mean with templated function. Do you mean a templated function in the base, which will be specified within each specific derived class?

Finally, after working a bit with the functions I would suggest following interfaces:
Set <- sets values
Calculate <- calculates something on the geometry and can update memory in the geometry
Get <- const function which cannot modify memory. However, Get would maybe not be required.

**Comment by pooyan-dadvand** [comment]
My concerns are:
- Set is used for flags everywhere
- Having Set and SetValue doing two different things is not easy to understand and naming is not helping either.

**Comment by tteschemacher** [comment]
I'd be happy to change the name. But I really think it'll be a big and important improvement.

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee discussed about this PR and there are some coments and concerns:
1. Set is used for flag and `Assign` or `CustomSetValue` would be better names. 
2. Calculate is const so it cannot change the geometry
3. `Calculate` and `Get` are very similar and in the element and conditions we omit the GetValueOnIntegrationPoint in favor of CalculateOnIntegrationPoint because it wasn't clear which one to use from outside.
4. Having Calculate modifying the geometry is not very natural and we had the same experience in the same usage of the Calculate in the elements

Considering above concern our proposal is:
1. Changing the set methods to `Assign`
2. Keep the `Calculate` constant and it only returns the asked value without modify the state of the geometry. (one can use Assign to change the geometry)
3. Do not add `Get` interface as they are similar to the `Calculate` and for a given variable (like `RADIUS`) one geometry may have it stored and another one may need a large calculation to give it. 

@philbucher would you please give us your opinion?

@KratosMultiphysics/dem does this proposal works for you?

**Comment by philbucher** [comment]
sounds good to me!

**Comment by tteschemacher** [comment]
I would be ok with the naming.

Thanks a lot @KratosMultiphysics/technical-committee . I can imagine that there was a decent discussion about this :P

**Comment by tteschemacher** [comment]
I changed it accordingly.

**Comment by maceligueta** [comment]
If I understood correctly, we will have this:
```
my_geometry.Assign(RADIUS, radius_to_be_set)
my_geometry.Calculate(RADIUS, recovered_radius)
```
I foresee that this will require a
`if (rVariable == RADIUS) { ... }`
inside. Is that correct?
Unless we limit the use of those two functions exclusively for RADIUS at compile time, then we would not verify the variable at runtime... I am here talking about some derived, optimized _Sphere_ geometry, not the one in the core.

**Comment by tteschemacher** [comment]
> 
> 
> If I understood correctly, we will have this:
> 
> ```
> my_geometry.Assign(RADIUS, radius_to_be_set)
> my_geometry.Calculate(RADIUS, recovered_radius)
> ```
> 
> I foresee that this will require a
> `if (rVariable == RADIUS) { ... }`
> inside. Is that correct?
> Unless we limit the use of those two functions exclusively for RADIUS at compile time. Then we would not verify the variable at runtime... I am here talking about some derived, optimized _Sphere_ geometry, not the one in the core.

I think I understand your point. It will be always a slice more expensive than having a specific interface. However, having a specific interface each time would be a bit an overkill, whereby, when ultimate speed is required this is of course essential.

In contrast, this interface will be much more efficient than using the dynamic memory via GetValue/SetValue. Requires search n the data base, dynamic memory, ... We can have many else if s to compensate this ;)

Your example is also maybe not the ultimate issue. Consider maybe an embedded object:
my_geometry.Assign(LOCAL_TANGENTS, local_tangents) // used to maybe update throughout the simulation, as some local tangents would be already apparent.
my_geometry.Calculate(GLOBAL_TANGENTS, global_tangents) // computed upon the element.

I think the if else can be speed up significantly with a switch, for an int comparison this is quite optimized.

**Comment by tteschemacher** [comment]
yo,

I changed the titel to fit it to the new necessities.

Shall we give it a go?

**Comment by maceligueta** [comment]
Ok from @KratosMultiphysics/dem . We will profile the cost of those `if`, eventually, for the DEM. The branch prediction may just vanish that cost.

**Comment by tteschemacher** [comment]
> Ok from @KratosMultiphysics/dem . We will profile the cost of those `if`, eventually, for the DEM. The branch prediction may just vanish that cost.

If you will profile it, I'd be interested to hear about the results.

**Comment by rubenzorrilla** [comment]
> yo,
> 
> I changed the titel to fit it to the new necessities.
> 
> Shall we give it a go?

@KratosMultiphysics/technical-committee agrees about the API enhancement and the current implementation.

What we want to discuss further is about the exposure of the `array_1d<double,2>`. While you have valid use cases for it, we don't want this to be spread. We would discuss about it in the next meeting.

**Comment by tteschemacher** [comment]
> > yo,
> > I changed the titel to fit it to the new necessities.
> > Shall we give it a go?
> 
> @KratosMultiphysics/technical-committee agrees about the API enhancement and the current implementation.
> 
> What we want to discuss further is about the exposure of the `array_1d<double,2>`. While you have valid use cases for it, we don't want this to be spread. We would discuss about it in the next meeting.

It's not an invasive spreading. But numerical methods come along with a massive set of parameters which have 2 components. If we want to have a clean and understandable code not good to fight against them.

**Review by rubenzorrilla — APPROVED** [review]
After a brief discussion in the @KratosMultiphysics/technical-committee , we agree that there are valid use cases for the `array_1d<double,2>`. However, for the moment these must stick to geometrical operations as that ones in here.

**Comment by philbucher** [comment]
>  However, for the moment these must stick to geometrical operations as that ones in here.

I.e. NO updates in the rest of the Code like `VariableUtils`

**Comment by tteschemacher** [comment]
Guys, this is like birthday and Christmas together :)

Thema for the review and the comprehensive discussion.

**Pull request merged**